### DISCUSSION TITLE
Unable to register websocket endpoint programatically

### DISCUSSION THREAD

**Discussion by masecla22** [quarkusio/quarkus#40576] [Q&A]
Hello!

I've been working on implementing the SignalR protocol in Quarkus.

I'm have been able to implement most of what I needed on the HTTP side, but I'm having issues deploying the websocket programatically.

I've implemented the following build steps

```java
@BuildStep
ReflectiveClassBuildItem markSignalRWebsocketForReflection() {
    return ReflectiveClassBuildItem.builder(SignalRWebsocket.class).methods().build();
}
```

```java
@BuildStep
@Record(ExecutionTime.STATIC_INIT)
void runWebsocketDeployment(
        List<SignalRHubBuildItem> signalHubs,
        BuildProducer<ReflectiveClassBuildItem> reflection,
        SignalRRecorder recorder) {
    for (SignalRHubBuildItem hub : signalHubs) {
        if (hub.getTransports().contains(TransportType.WEB_SOCKETS))
            try {
                recorder.deployWebsocket(hub.getPath(), hub.getClassName());
            } catch (DeploymentException e) {
                log.error("Failed to deploy websocket for hub at " + hub.getPath(), e);
            }
    }
}
```

And then I deploy my websockets within the recorder as such
```java
@Recorder
public class SignalRRecorder {
    public void addHub(String hubPath, String hubClass) {
        Arc.container().instance(SignalRConnectorManager.class).get().addHub(hubPath, hubClass);
    }

    public void deployWebsocket(String hubPath, String hubClass) throws DeploymentException {
        ServerContainer container = (ServerContainer) ContainerProvider.getWebSocketContainer();
        container.addEndpoint(ServerEndpointConfig.Builder.create(SignalRWebsocket.class, hubPath).build());
        System.out.println("Websocket deployed for path: " + hubPath + " with class: " + hubClass);
    }
}
```

For some reason, although I get the logger message
```
UT026005: Adding programmatic server endpoint class nl.rug.digitallab.signalr.extension.runtime.websockets.SignalRWebsocket for path /bye/hub
```

If I try to connect to the websocket instance, I constantly get a 404. Is there anything I forgot to implement or register?

I've also tried consuming ServerWebSocketContainerBuildItem, but no matter where I put it, I get a null value for it.

**Comment by phillip-kruger** [comment]
What is the underlying HTTP technology? Vertx? Servlet? Something else?

**Reply by masecla22** [reply]
I'm using the default Quarkus extensions, so it's Vert.X as far as I can tell. 

![image](https://github.com/quarkusio/quarkus/assets/21177004/6ba605b5-3d62-4c97-881d-d59e9d408ce6)

Here are my extensions

**Reply by cescoffier** [reply]
I know it's confusing but the default websocket support is based on undertow (running on Vert.x). 

We recently released websocket-next, a new approach based on Vertx directly.

In your case, however, I'm not sure it is necessary. You can register a route on the main router (@Inject Router router) and upgrade to websocket in the request handler.

**Comment by phillip-kruger** [comment]
So SignalelRWebsocket is a vertx websocket? The 404 indicate that the vertx router does not know about it.

**Reply by masecla22** [reply]
I annotated it with the methods, though I might have misunderstood how websockets work in Quarkus

```java

public class SignalRWebsocket {
    @OnOpen
    public void onOpen(Session session) {
        System.out.println("Websocket opened");
    }

    @OnClose
    public void onClose(Session session) {
        System.out.println("Websocket closed");
    }

    @OnMessage
    public void onMessage(String message) {
        System.out.println("Message received: " + message);
    }
}
```

I'm trying to hook directly into the Quarkus extension rather than the underlying container.

Also, is there a reason ServerWebSocketContainerBuildItem would be null within my BuildStep parameters?

**Reply by phillip-kruger** [reply]
I would say it's better to use Vertx directly, rather than using the Quarkus websocket extension. (less dependencies to pull in). I assume you currently have an dependency on quarkus-websocket ? Here is an example (the Dev UI) that use vertx websocket: https://github.com/quarkusio/quarkus/blob/7c172ad413e55f58c50ba841e7a732a11d01daed/extensions/vertx-http/deployment/src/main/java/io/quarkus/devui/deployment/DevUIProcessor.java#L151

and

https://github.com/quarkusio/quarkus/blob/7c172ad413e55f58c50ba841e7a732a11d01daed/extensions/vertx-http/runtime/src/main/java/io/quarkus/devui/runtime/DevUIRecorder.java#L74

and

https://github.com/quarkusio/quarkus/blob/main/extensions/vertx-http/runtime/src/main/java/io/quarkus/devui/runtime/DevUIWebSocket.java#L16

**Reply by masecla22** [reply]
Yes, I currently have as dependencies the quarkus-websocket, but I'll see if I can implement the socket directly.

Is there any particular reason the Quarkus extension wouldn't pick up on my programmatic endpoints?

Also, another thing I noticed (I will open a separate thread if needed), when I'm running my code without the extension, in the dev UI the endpoints tab works fine, but when I add my extension as a dependency, it gets stuck on the Fetching information phase. 

Is that expected? The Rest extension is still able to get endpoint information to show the score of each endpoint.

**Reply by phillip-kruger** [reply]
> Is there any particular reason the Quarkus extension wouldn't pick up on my programmatic endpoints?

When using quarkus-websocket ? I am not sure, but you might have to do something in your extension to let the quarkus-websocket your exist (if that is even possible). Again, rather use vertx.

W.r.t your dev ui problem. Are you saying, when adding your extension the dev ui does not work anymore ? That should not happen, if you can share some reproducer I can have a look

**Reply by masecla22** [reply]
[signalr-extension.zip](https://github.com/quarkusio/quarkus/files/15284369/signalr-extension.zip)

Here, is my reproducer.

Simply run test.sh, it should compile the extension first (to your local maven repository) and will then start the quarkus application.

Removing the annotations from GreetingResource and GreetingResource2 doesn't fix the issue, but fully removing the extension does.

I've also tried commenting out every single build step and running it like that, but it didn't seem to help.

Let me know if you can reproduce the issue!

**Reply by masecla22** [reply]
For clarification, this is the part of the Dev UI I'm referring to:
![image](https://github.com/quarkusio/quarkus/assets/21177004/1caf354e-427a-4eaf-a4c4-c891ede140e9)

**Reply by phillip-kruger** [reply]
I can reproduce the Dev UI issue. It might be related. (It's todo with the 404 handling) . I'll have a deeper look tomorrow. Let me know if this still happens if you convert your extension to use vertx

**Reply by masecla22** [reply]
Thank you very much, I'll start porting to vertx and see what results I get!

**Reply by phillip-kruger** [reply]
I see you are also using servlet for HTTP endpoints. You should do the same for HTTP, drop Undertow and use vertx.

**Reply by masecla22** [reply]
Got it, thanks!
